mlir/lib/Conversion/QuantumToMQTOpt/QuantumToMQTOpt.cpp

```diff
@@ -0,0 +1,476 @@
+    // needed.
+    addConversion([](Type type) { return type; });
+
+    // Convert source QuregType to target QubitRegisterType
+    addConversion([ctx](catalyst::quantum::QuregType /*type*/) -> Type {
+      return ::mqt::ir::opt::QubitRegisterType::get(ctx);
+    });
+
+    // Convert source QubitType to target QubitType
+    addConversion([ctx](catalyst::quantum::QubitType /*type*/) -> Type {
+      return ::mqt::ir::opt::QubitType::get(ctx);
+    });
+  }
+};
+
+struct ConvertQuantumAlloc
+    : public OpConversionPattern<catalyst::quantum::AllocOp> {
+
+  ConvertQuantumAlloc(const TypeConverter& typeConverter, MLIRContext* context)
+      : OpConversionPattern<catalyst::quantum::AllocOp>(typeConverter,
+                                                        context) {}
+
+  LogicalResult
+  matchAndRewrite(catalyst::quantum::AllocOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter& rewriter) const override {
+
+    // Extract operand(s) and attribute(s)
+    auto const nQubitsValue = adaptor.getNqubits();
+    auto const nQubitsIntegerAttr = adaptor.getNqubitsAttrAttr();
+
+    // Prepare the result type(s)
+    auto resultType =
+        ::mqt::ir::opt::QubitRegisterType::get(rewriter.getContext());
+
+    // Create the new operation
+    auto mqtoptOp = rewriter.create<::mqt::ir::opt::AllocOp>(
+        op.getLoc(), resultType, nQubitsValue, nQubitsIntegerAttr);
+
+    // Get the result of the new operation, which represents the qubit register
+    auto trgtQreg = mqtoptOp->getResult(0);
+
+    // Collect the users of the original operation to update their operands
+    std::vector<mlir::Operation*> users(op->getUsers().begin(),
+                                        op->getUsers().end());
+
+    // Iterate over the users in reverse order
+    for (auto* user : llvm::reverse(users)) {
+      // Registers should only be used in Extract, Insert or Dealloc operations
+      if (mlir::isa<catalyst::quantum::ExtractOp>(user) ||
+          mlir::isa<catalyst::quantum::InsertOp>(user) ||
+          mlir::isa<catalyst::quantum::DeallocOp>(user)) {
+        // Update the operand of the user operation to the new qubit register
+        user->setOperand(0, trgtQreg);
+      }
+    }
+
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+struct ConvertQuantumDealloc
+    : public OpConversionPattern<catalyst::quantum::DeallocOp> {
+
+  ConvertQuantumDealloc(const TypeConverter& typeConverter,
+                        MLIRContext* context)
+      : OpConversionPattern<catalyst::quantum::DeallocOp>(typeConverter,
+                                                          context) {}
+
+  LogicalResult
+  matchAndRewrite(catalyst::quantum::DeallocOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter& rewriter) const override {
+
+    // Extract operand(s)
+    auto qregValue = adaptor.getQreg();
+
+    // Prepare the result type(s)
+    auto resultTypes = ::mlir::TypeRange({});
```

Comment:
Seems redundant. Can't this simply be inlined below with `{}` for the function parameter?